### PR TITLE
[exporter/fileexporter] add periodic flushing

### LINKED ISSUE #18251: [exporter/file] Add buffer flush interval
### Component(s)

exporter/file

### Is your feature request related to a problem? Please describe.

The observability team at my company maintains a custom version of the collector. The custom collector includes custom receivers, processors, and exporters for all OpenTelemetry data types. The file exporter is used to verify the full data flow by comparing known inputs to their expected output payloads.
In v0.70.0, the file exporter moved to a buffered writer. This breaks the tests because we cannot reliably ensure that the payloads have been written to disk when the check happens. Having the buffered output as the default makes sense; however, we'd also like a mechanism for turning off the buffering for these types of use cases.

### Describe the solution you'd like

Add a parameter to turn off the buffering within the file exporter.

### Describe alternatives you've considered

We've discussed using the OpenTelemetry testbed or using a real backend for collecting the data. However, both of these solutions would significantly complicate the testing environment. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by peterzandbergen**
**Description:** <Describe what has changed.>

Added flushing to the fileExporter to ease live debugging.
Added flush_interval parameter to the Config struct. 
The value of flush_interval should be 1 or higher. 
If not flushing is disabled.
Flushing occures every flush_interval seconds.

**Link to tracking Issue:**
Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18251

**Testing:** <Describe what testing was performed and which tests were added.>
Tests were added to config_test.go, file_exporter_test.go. These passed.
Test were also performed by running the otelcontrib program. 

**Documentation:** <Describe the documentation added.>
The README.md file has been updated with the new flush_interval setting.

### CODE REVIEW

**Comment by djaglowski** [comment]
@peterzandbergen, thanks for contributing this. Please sign the CLA.

We can use https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18251 as the tracking issue since this need was noted there as well.

**Comment by peterzandbergen** [comment]
CLA is signed

File: exporter/fileexporter/file_exporter.go
Comment (djaglowski) [inline exporter/fileexporter/file_exporter.go:162]: I would prefer that we determine here whether or not to call `startFlusher` and that we do so by comparing to the user specified value. I think this keeps the business logic clearer.

```go
if e.flushInterval > 0 {
  e.startFlusher()
}
```

File: exporter/fileexporter/file_exporter.go
Comment (djaglowski) [inline exporter/fileexporter/file_exporter.go:123]: Rather than pass the exporter into the function, let's make this a receiver method.
```suggestion
func (e *fileExporter) startFlusher() {
```

File: exporter/fileexporter/file_exporter.go
Comment (djaglowski) [inline exporter/fileexporter/file_exporter.go:152]:
```suggestion
func (e *fileExporter) stopFlusher() {
```

File: exporter/fileexporter/file_exporter.go
Comment (djaglowski) [inline exporter/fileexporter/file_exporter.go:169]:
```suggestion
	if e.flushTicker != nil {
		stopFlusher(e)
	}
```

File: exporter/fileexporter/testdata/config.yaml
Comment (djaglowski) [inline exporter/fileexporter/testdata/config.yaml:45]:
```suggestion
  flush_interval: 5

```

File: exporter/fileexporter/config.go
Comment (djaglowski) [inline exporter/fileexporter/config.go:49]: Typically we will use a `time.Duration` for measures of time. This allows users to pick more aggressive intervals if they want, or easily use very long ones. e.g. `100ns`, `2h`

File: exporter/fileexporter/file_exporter.go
Comment (atingchen) [inline exporter/fileexporter/file_exporter.go:132]: When `Start` calls `startFlusher`, I don't think there will be a situation where e.flushTicker != nil. As djaglowski said, we should determine whether to call the `startFlusher`  when the `Start` is called. So, can these two if statements be deleted?

File: exporter/fileexporter/file_exporter.go
Comment (peterzandbergen) [reply exporter/fileexporter/file_exporter.go:123]: Understood

File: exporter/fileexporter/file_exporter.go
Comment (peterzandbergen) [reply exporter/fileexporter/file_exporter.go:162]: Changed

File: exporter/fileexporter/file_exporter.go
Comment (peterzandbergen) [reply exporter/fileexporter/file_exporter.go:152]: Changed

File: exporter/fileexporter/file_exporter.go
Comment (peterzandbergen) [reply exporter/fileexporter/file_exporter.go:169]: Changed

File: exporter/fileexporter/file_exporter.go
Comment (peterzandbergen) [reply exporter/fileexporter/file_exporter.go:132]: You have a point. I have changed it.

File: exporter/fileexporter/testdata/config.yaml
Comment (peterzandbergen) [reply exporter/fileexporter/testdata/config.yaml:45]: I have added test cases for different units for flush_interval.

File: exporter/fileexporter/config.go
Comment (peterzandbergen) [reply exporter/fileexporter/config.go:49]: I  have changed it to Duration.

**Comment by peterzandbergen** [comment]
Hi Daniel and Iook (I hope this is first name),
Thanks for your comments and suggestions for improvement. I have incorporate them all.
I have also improved the way the goroutine stops.

Changes
- Change flush_interval to time.Duration
- Change startFlusher to a receiver method
- Moved stopFlusher logic to Stop method
- Move test on flush_interval to Start method
- Remove superfluous tests
- Add stopFlusher channel to ensure that the flushing goroutine
  stops correctly

**Comment by peterzandbergen** [comment]
Hi, can you let me know if I need to perform more changes? Cheers

**Comment by djaglowski** [comment]
Thanks for making this contribution @peterzandbergen. The code looks good to me and I've verified manually that it works as expected. 

There are two things I think we need to consider yet. 
1. Should we enable this functionality by default? I think probably yes, as this component is commonly used for manual testing and the lack of automatic flushing often causes confusion. I don't see any notable downsides to having an unaggressive flush interval. I think `1s` would be reasonable. WDYT? Perhaps @atingchen, do you have an opinion on this?
2.  It's not clear to me that any of the tests are actually validating this functionality. I think we should have a test that ensures the flush occurs. Apologies if I missed this but I couldn't find it.

**Comment by peterzandbergen** [comment]
Hi @djaglowski,
thanks for your comment. 

ad 1) I agree that a default flush interval makes sense. When I started using the fileexporter I thought I was doing something wrong when I saw no output. A default of 1 second sounds reasonable.

ad 2) I will think of a test to validate if flushing works.

**Comment by atingchen** [comment]
> 1. Should we enable this functionality by default? I think probably yes, as this component is commonly used for manual testing and the lack of automatic flushing often causes confusion. I don't see any notable downsides to having an unaggressive flush interval. I think `1s` would be reasonable. WDYT? Perhaps @atingchen, do you have an opinion on this?

I think this sounds like a good idea.

**Review by djaglowski — APPROVED** [review]

**Comment by djaglowski** [comment]
@peterzandbergen, I'll be there and look forward to connecting.

**Comment by atingchen** [comment]
@peterzandbergen Thank you very much for your invitation. Unfortunately, I won't be able to make it to the event.Hope you have a happy memory.

**Comment by peterzandbergen** [comment]
@atingchen Hi Luke, it was great working together. Let's do it again sometime.

**Comment by peterzandbergen** [comment]
I am adding commits to fix race errors in testing.

**Comment by peterzandbergen** [comment]
@djaglowski Hi, I see that you have approved it. I is now waiting for another reviewer. Is there something I need to provide for it to be approved?
I am asking because I am not familiar with the whole process.
Cheers

BTW, are you attending KubeCon as a speaker or does your company have a booth?

File: exporter/fileexporter/file_exporter.go
Comment (djaglowski) [inline exporter/fileexporter/file_exporter.go:185]:
```suggestion
```

**Comment by djaglowski** [comment]
@peterzandbergen, thanks for resolving the race condition. There was one more lint issue but I believe I've resolved it. I'll reach out on CNCF Slack re KubeCon.

@atingchen, do you want to give this a final review?

**Comment by peterzandbergen** [comment]
@djaglowski @atingchen Dan and Luke, 
I want to thank you for your cooperation and patience in finishing this PR. 
I was not familiar with the way the project was set up, especially I needed to learn about the different options for make for lint and test and to localize this to the exporter/fileexporter directory. 
All in all it was a good experience and it encourages me to contribute more. 
Again, thanks.

**Comment by peterzandbergen** [comment]
One last thing, I hope. There is no check on a negative duration for flush_interval, it will result in panic when the ticker is created. I suggest to return an error in this case and document it.

**Comment by atingchen** [comment]
> @peterzandbergen, thanks for resolving the race condition. There was one more lint issue but I believe I've resolved it. I'll reach out on CNCF Slack re KubeCon.
> 
> @atingchen, do you want to give this a final review?

It looks good to me, but I'm not authorized to approved it.

**Comment by EdmundMartin** [comment]
Is anyone actively looking at this to provide the second review? Found the fact that the file exporter doesn't periodically flush to disk quite surprising and this PR would be a good fit for my use case. Would love to see this get merged in!

Is there anything that can be done to get this change over the line?

**Comment by djaglowski** [comment]
Thanks for the contribution @peterzandbergen!

**Pull request merged**
